File: smart_attendance/attendance/qr_utils.py
"""
QR Code Utilities
=================
Implements the HMAC-SHA256 QR payload generation and verification
described in the research proposal (Section 3.3 & 3.5).

QR Payload Format (pipe-delimited, signed with HMAC-SHA256):
    <session_id>|<course_code>|<expiry_unix>|<hmac_hex>

The HMAC key is the session_secret stored only on the server and
transmitted to the Lecturer device at session-creation time.
Students never receive the secret; they receive only the signed
QR payload — ensuring they cannot forge a valid code.
"""
import hashlib
import hmac
import secrets
import time


def generate_session_secret() -> str:
    """Generate a cryptographically secure 64-byte hex session secret."""
    return secrets.token_hex(64)


def build_payload_base(session_id: str, course_code: str, expiry_unix: int) -> str:
    """Construct the canonical string that is signed."""
    return f"{session_id}|{course_code}|{expiry_unix}"


def compute_hmac(payload_base: str, secret: str) -> str:
    """Return HMAC-SHA256 hex digest of `payload_base` using `secret`."""
    return hmac.new(
        secret.encode("utf-8"),
        payload_base.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def generate_qr_payload(
    session_id: str,
    course_code: str,
    secret: str,
    validity_seconds: int = 900,
) -> dict:
    """
    Build a complete, signed QR payload dict.

    Returns
    -------
    {
        "payload": "<base>|<hmac>",   # The string encoded into the QR image
        "expiry_unix": int,
        "hmac": str,
    }
    """
    expiry_unix = int(time.time()) + validity_seconds
    payload_base = build_payload_base(session_id, course_code, expiry_unix)
    signature = compute_hmac(payload_base, secret)
    return {
        "payload": f"{payload_base}|{signature}",
        "expiry_unix": expiry_unix,
        "hmac": signature,
    }


def parse_qr_payload(raw_payload: str) -> dict | None:
    """
    Parse a raw QR string into components.

    Returns None if the format is invalid.
    """
    parts = raw_payload.split("|")
    if len(parts) != 4:
        return None
    session_id, course_code, expiry_unix_str, signature = parts
    try:
        expiry_unix = int(expiry_unix_str)
    except ValueError:
        return None
    return {
        "session_id": session_id,
        "course_code": course_code,
        "expiry_unix": expiry_unix,
        "signature": signature,
    }


def verify_qr_payload(
    raw_payload: str,
    secret: str,
    clock_skew_tolerance: int = 300,
) -> tuple[bool, str]:
    """
    Fully verify a QR payload.

    Checks
    ------
    1. Correct format (4 pipe-delimited parts).
    2. HMAC-SHA256 signature is valid.
    3. Payload has not expired (with clock_skew_tolerance).

    Returns
    -------
    (True, "ok") on success or (False, "<reason>") on failure.
    """
    parsed = parse_qr_payload(raw_payload)
    if parsed is None:
        return False, "invalid_format"

    payload_base = build_payload_base(
        parsed["session_id"],
        parsed["course_code"],
        parsed["expiry_unix"],
    )
    expected_hmac = compute_hmac(payload_base, secret)

    if not hmac.compare_digest(expected_hmac, parsed["signature"]):
        return False, "invalid_hmac"

    now = int(time.time())
    if now > parsed["expiry_unix"] + clock_skew_tolerance:
        return False, "expired"

    return True, "ok"

File: smart_attendance_flutter/lib/core/utils/qr_utils.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// QR Payload Format:  session_id|course_code|expiry_unix|hmac_sha256
class QrUtils {
  QrUtils._();

  static String computeHmac(String payloadBase, String secret) {
    final key = utf8.encode(secret);
    final bytes = utf8.encode(payloadBase);
    final hmac = Hmac(sha256, key);
    return hmac.convert(bytes).toString();
  }

  static String buildPayloadBase(String sessionId, String courseCode, int expiryUnix) =>
      '$sessionId|$courseCode|$expiryUnix';

  static String generateQrPayload({
    required String sessionId,
    required String courseCode,
    required String secret,
    int validitySeconds = 900,
  }) {
    final expiry = (DateTime.now().millisecondsSinceEpoch ~/ 1000) + validitySeconds;
    final base = buildPayloadBase(sessionId, courseCode, expiry);
    final sig = computeHmac(base, secret);
    return '$base|$sig';
  }

  static QrParseResult? parseQrPayload(String raw) {
    final parts = raw.split('|');
    if (parts.length != 4) return null;
    final expiry = int.tryParse(parts[2]);
    if (expiry == null) return null;
    return QrParseResult(
      sessionId: parts[0],
      courseCode: parts[1],
      expiryUnix: expiry,
      signature: parts[3],
    );
  }

  static QrVerifyResult verifyQrPayload(
    String raw,
    String secret, {
    int clockSkewTolerance = 300,
  }) {
    final parsed = parseQrPayload(raw);
    if (parsed == null) return QrVerifyResult(ok: false, reason: 'invalid_format');

    final base = buildPayloadBase(parsed.sessionId, parsed.courseCode, parsed.expiryUnix);
    final expected = computeHmac(base, secret);

    if (expected != parsed.signature) {
      return QrVerifyResult(ok: false, reason: 'invalid_hmac');
    }

    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    if (now > parsed.expiryUnix + clockSkewTolerance) {
      return QrVerifyResult(ok: false, reason: 'expired');
    }

    return QrVerifyResult(ok: true, reason: 'ok', parsed: parsed);
  }
}

class QrParseResult {
  final String sessionId;
  final String courseCode;
  final int expiryUnix;
  final String signature;

  const QrParseResult({
    required this.sessionId,
    required this.courseCode,
    required this.expiryUnix,
    required this.signature,
  });
}

class QrVerifyResult {
  final bool ok;
  final String reason;
  final QrParseResult? parsed;
  const QrVerifyResult({required this.ok, required this.reason, this.parsed});
}
